addons/sale_timesheet/models/account_move.py

```diff
@@ -78,3 +78,10 @@ def _timesheet_domain_get_invoiced_lines(self, sale_line_delivery):
             ('project_id', '!=', False),
             '|', ('timesheet_invoice_id', '=', False), ('timesheet_invoice_id.state', '=', 'cancel')
         ]
+
+    def unlink(self):
+        for line in self.filtered(lambda i: i.move_id.move_type == 'out_invoice' and i.move_id.state == 'draft'):
+            sale_line_delivery = line.sale_line_ids.filtered(lambda sol: sol.product_id.invoice_policy == 'delivery' and sol.product_id.service_type == 'timesheet')
+            timesheets = line.move_id.timesheet_ids.filtered(lambda aal: aal.so_line in sale_line_delivery)
+            timesheets.write({'timesheet_invoice_id': False})
```

Comment:
It is better to avoid calling a filtered (which is actually a loop) in a for loop definition as it creates a loop before your loop (in that case it is better to check your condition inside the loop).
It is also better to try not to call a filtered inside another loop 🙂 as it create a nested loop.
I haven't checked the code I wrote with @xavierbol but it should be something close to this:
```suggestion
    def unlink(self):
        timesheets = self.env['account.analytic.line']
        move_line_read_group = self.env['account.move.line'].read_group([
            ('move_id.move_type', '=', 'out_invoice'),
            ('move_id.state', '=', 'draft'),
            ('sale_line_ids.product_id.invoice_policy', '=', 'delivery'),
            ('sale_line_ids.product_id.service_type', '=', 'timesheet')],
            ['move_id', 'sale_line_ids:array_agg(sale_line_ids)'],
            ['move_id'])
        sale_line_ids_per_move = {res['move_id'][0]: res['sale_line_ids'] for res in move_line_read_group}
        timesheet_read_group = self.env['account.analytic.line'].read_group([
            ('timesheet_invoice_id.move_type', '=', 'out_invoice'),
            ('timesheet_invoice_id.state', '=', 'draft')], 
            ['timesheet_invoice_id', 'so_line', 'ids:array_agg(id)'], 
            ['timesheet_invoice_id', 'so_line'], 
            lazy=False)
        timesheet_ids = []
        for res in timesheet_read_group:
            move_id = res['timesheet_invoice_id'][0]
            if res['so_line'][0] in sale_line_ids_per_move[move_id]:
                timesheet_ids += res['ids']
        self.env['account.analytic.line'].browse(timesheet_ids).write({'timesheet_invoice_id': False})        
```
Can you check it and try something similar?
Thanks 🙂